async_work_promise/napi/binding.c

```diff
@@ -0,0 +1,171 @@
+#include <assert.h>
+#include <stdlib.h>
+#define NAPI_EXPERIMENTAL
+#include <node_api.h>
+
+// Limit ourselves to this many primes, starting at 2
+#define PRIME_COUNT 10
+
+typedef struct linked_box_s linked_box_t;
+struct linked_box_s {
+  linked_box_t* prev;
+  int the_prime;
+};
+
+typedef struct {
+  napi_async_work work;
+  napi_deferred deferred;
+  linked_box_t* box_head;
+} AddonData;
+
+// This function runs on a worker thread. It has no access to the JavaScript.
+static void ExecuteWork(napi_env env, void* data) {
+  AddonData* addon_data = (AddonData*)data;
+  int idx_inner, idx_outer;
+  int prime_count = 0;
+
+  // Find the first 1000 prime numbers using an extremely inefficient algorithm.
+  for (idx_outer = 2; prime_count < PRIME_COUNT; idx_outer++) {
+    for (idx_inner = 2; idx_inner < idx_outer; idx_inner++) {
+      if (idx_outer % idx_inner == 0) {
+        break;
+      }
+    }
+    if (idx_inner < idx_outer) {
+      continue;
+    }
+
+    ++prime_count;
+    // Save the prime number to the heap. The JavaScript marshaller (CallJs)
+    // will free this item after having sent it to JavaScript.
+    linked_box_t* the_box = malloc(sizeof(linked_box_t));
+    the_box->the_prime = idx_outer;
+    the_box->prev = addon_data->box_head;
+    addon_data->box_head = the_box;
+  }
+}
+
+// This function runs on the main thread after `ExecuteWork` exits.
+static void WorkComplete(napi_env env, napi_status status, void* data) {
+  AddonData* addon_data = (AddonData*)data;
+  linked_box_t* the_box = addon_data->box_head;
+  uint32_t idx = 0;
+  napi_value array;
+  assert(napi_create_array(env, &array) == napi_ok);
+
+  while (the_box != NULL) {
+    napi_value js_the_prime;
+
+    // Convert the integer to a napi_value.
+    assert(napi_create_int32(env, the_box->the_prime, &js_the_prime) == napi_ok);
+
+    // Store the js value in the array.
+    assert(napi_set_element(env, array, idx, js_the_prime) == napi_ok);
+
+    idx++;
+    linked_box_t* swap = the_box;
+    the_box = the_box->prev;
+    free(swap);
+  }
+  assert(napi_resolve_deferred(env, addon_data->deferred, array) == napi_ok);
```

Comment:
Let's change these to
```C
napi_status status = napi_call();
assert(status == napi_ok);
```

Otherwise the actual N-API calls will only be compiled in debug mode.

Reply:
Yes, you are right. I'll update the examples.